### PR TITLE
Corrected testcase CustomResourceBackupAndRestore

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PB-5263
**Special notes for your reviewer**:

log :S3:  https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3871/consoleFull
NFS : https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3881/console

After change : https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3898/consoleFull
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3910/consoleFull

### CODE REVIEW

File: tests/backup/backup_restore_basic_test.go [hunk 1]
```diff
@@ -251,25 +253,64 @@ var _ = Describe("{CustomResourceBackupAndRestore}", func() {
...
+				mutex.Lock()
+				go func(restoreName string, nsMapping map[string]string) {
+					defer GinkgoRecover()
+					defer wg.Done()
+					log.InfoD("Validating restore [%s]", restoreName)
+					var expectedRestoredAppContextList []*scheduler.Context
+					for _, context := range restoreContextMap[restoreName] {
+						expectedRestoredAppContext, err := CloneAppContextAndTransformWithMappings(context, nsMapping, make(map[string]string), true)
+						if err != nil {
+							errChan <- err
```
Comment (kphalgun-px) [inline tests/backup/backup_restore_basic_test.go:292]:
```go
errChan <- err
break // continue
```

**Review by kphalgun-px — APPROVED** [review]
LGTM

File: tests/backup/backup_restore_basic_test.go [hunk 2]
```diff
@@ -251,25 +253,64 @@ var _ = Describe("{CustomResourceBackupAndRestore}", func() {
...
+			dash.VerifyFatal(len(errors), 0, fmt.Sprintf("Errors generated while validating restores of individual backups -\n%s", strings.Join(errors, "}\n{")))
+		})
+
 		Step("Compare PVCs on both namespaces", func() {
 			log.InfoD("Compare PVCs on both namespaces")
-			for _, namespace := range bkpNamespaces {
-				pvcs, _ := core.Instance().GetPersistentVolumeClaims(namespace, labelSelectors)
-				restoreNamespace := fmt.Sprintf("%s-%s", namespace, "restored")
-				restoredPvcs, _ := core.Instance().GetPersistentVolumeClaims(restoreNamespace, labelSelectors)
+			for sourceNamespace, restoredNamespace := range namespaceMapping {
```
Comment (kphalgun-px) [inline tests/backup/backup_restore_basic_test.go:311]: 👍

File: tests/backup/backup_restore_basic_test.go [hunk 3]
```diff
@@ -251,25 +253,64 @@ var _ = Describe("{CustomResourceBackupAndRestore}", func() {
...
+				restoreNamespaceMappingMap[restoreName] = namespaceMapping
 				appContextsToBackup := FilterAppContextsByNamespace(scheduledAppContexts, namespaceList)
-				err = CreateRestoreWithValidation(ctx, restoreName, backupName, namespaceMapping, make(map[string]string), SourceClusterName, orgID, appContextsToBackup)
+				restoreContextMap[restoreName] = appContextsToBackup
+				err = CreateRestore(restoreName, backupName, namespaceMapping, SourceClusterName, orgID, ctx, make(map[string]string))
 				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating restore: %s from backup: %s", restoreName, backupName))
 			}
 		})
 
+		Step("Validating custom resource backup restores", func() {
```
Comment (trenukarya-px) [inline tests/backup/backup_restore_basic_test.go:272]: @snigdha-px Can we make generic library ? Cc: @mkoppal-px @atrivedi-px

File: tests/backup/backup_helper.go [hunk 4]
```diff
@@ -2271,7 +2271,22 @@ func ValidateRestore(ctx context.Context, restoreName string, orgID string, expe
...
-
+		// This part is added when we have taken backup of custom resources and the restored namespace will not have all the resource as that of
+		// source namespace, so modifying the expectedRestoredAppContext to have only the resources which are present in the restored namespace
+		if len(resourceTypesFilter) != 0 {
+			newCtxAppSpecList := make([]interface{}, 0)
+			for _, spec := range expectedRestoredAppContext.App.SpecList {
+				val := reflect.ValueOf(spec)
+				if val.Kind() == reflect.Struct && val.FieldByName("Kind").IsValid() {
+					kindField := val.FieldByName("Kind")
+					if IsPresent(kindField, resourceTypesFilter) {
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_helper.go:2282]: done

File: tests/backup/backup_restore_basic_test.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/backup_restore_basic_test.go:292]: done

**Review by ak-px — APPROVED** [review]
LGTM

File: tests/backup/backup_restore_basic_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_restore_basic_test.go:272]: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3896/

Comment (snigdha-px) [reply tests/backup/backup_restore_basic_test.go:272]: done

**Review by trenukarya-px — APPROVED** [review]

File: tests/backup/backup_helper.go [hunk 5]
```diff
@@ -6545,3 +6545,47 @@ func SuspendAndDeleteSchedule(backupScheduleName string, schedulePolicyName stri
 	}
 	return nil
 }
+
+// ValidateCustomResourceRestore validates restore taken of custom resource
+func ValidateCustomResourceRestore(ctx context.Context, orgID string, resourceTypesFilter []string, restoreNamespaceScMappingMap map[string][]map[string]string, restoreContextMap map[string][]*scheduler.Context) error {
+	errChan := make(chan error, 100)
+	var errList []error
+	var wg sync.WaitGroup
+	for restoreName, nsScMapping := range restoreNamespaceScMappingMap {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:6554]: Add a sized check for `nsScMapping` before triggering the go routine

Comment (snigdha-px) [reply tests/backup/backup_helper.go:6573]: done

Comment (snigdha-px) [reply tests/backup/backup_helper.go:6573]: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3910/consoleFull

File: tests/backup/backup_helper.go [hunk 6]
```diff
@@ -6545,3 +6564,51 @@ func SuspendAndDeleteSchedule(backupScheduleName string, schedulePolicyName stri
 	}
 	return nil
 }
+
+// ValidateCustomResourceRestore validates restore taken of custom resource
+func ValidateCustomResourceRestore(ctx context.Context, orgID string, resourceTypesFilter []string, restoreNamespaceScMappingMap map[string][]map[string]string, restoreContextMap map[string][]*scheduler.Context) error {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:6569]: Split the map `restoreNamespaceScMappingMap` into `restoreNamespaceMap` and `restoreScMap`

Comment (mkoppal-px) [reply tests/backup/backup_helper.go:6569]: Change the name of the func to `ValidateCustomResourceRestores`

Comment (snigdha-px) [reply tests/backup/backup_helper.go:6569]: done

Comment (snigdha-px) [reply tests/backup/backup_helper.go]: https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/3912/consoleFull

File: tests/backup/backup_helper.go [hunk 7]
```diff
@@ -6545,3 +6564,72 @@ func SuspendAndDeleteSchedule(backupScheduleName string, schedulePolicyName stri
...
+			var expectedRestoredAppContextList []*scheduler.Context
+			for _, context := range contexts {
+				expectedRestoredAppContext, err := CloneAppContextAndTransformWithMappings(context, nsMapping, scMapping, true)
+				if err != nil {
+					errChan <- err
+					return
+				}
+				expectedRestoredAppContextList = append(expectedRestoredAppContextList, expectedRestoredAppContext)
+			}
+			err := ValidateRestore(ctx, restoreName, orgID, expectedRestoredAppContextList, resourceTypesFilter)
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:6613]: Are we logging anywhere about the resource type we are validating here?

**Review by mkoppal-px — APPROVED** [review]
LGTM except for one minor comment

File: tests/backup/backup_helper.go (hunk 7 above)
Comment (snigdha-px) [reply tests/backup/backup_helper.go]: done

**Pull request merged**